assets/lambda@edge/LambdaOriginRequestIamAuth.ts

```diff
@@ -0,0 +1,113 @@
+import { Sha256 } from '@aws-crypto/sha256-js';
+import { SignatureV4 } from '@aws-sdk/signature-v4';
+import type { CloudFrontHeaders, CloudFrontRequest, CloudFrontRequestHandler } from 'aws-lambda';
+import { fixHostHeader, handleS3Request } from './common';
+
+/**
+ * This fixes the "host" header to be the host of the origin.
+ * The origin is the lambda server function URL.
+ * If we don't provide its expected "host", it will not know how to route the request.
+ */
+export const handler: CloudFrontRequestHandler = async (event) => {
+  const request = event.Records[0].cf.request;
+  // console.log('request', JSON.stringify(request, null, 2));
+
+  handleS3Request(request);
+  fixHostHeader(request);
+  await signRequest(request);
+  // console.log(JSON.stringify(request), null, 2);
+  return request;
+};
+
+let sigv4: SignatureV4;
+
+/**
+ * When `NextjsDistributionProps.functionUrlAuthType` is set to
+ * `lambda.FunctionUrlAuthType.AWS_IAM` we need to sign the `CloudFrontRequest`s
+ * with AWS IAM SigV4 so that CloudFront can invoke the Nextjs server and image
+ * optimization functions via function URLs. When configured, this lambda@edge
+ * function has the permission, lambda:InvokeFunctionUrl, to invoke both
+ * functions.
+ * @link https://medium.com/@dario_26152/restrict-access-to-lambda-functionurl-to-cloudfront-using-aws-iam-988583834705
+ */
+async function signRequest(request: CloudFrontRequest) {
+  if (!sigv4) {
+    sigv4 = getSigV4();
+  }
+  const headerBag = cfHeadersToHeaderBag(request.headers);
+  // don't sign x-forwarded-for b/c it changes from hop to hop
+  delete headerBag['x-forwarded-for'];
+  let body: string | undefined;
+  if (request.body?.data) {
+    body = Buffer.from(request.body.data, 'base64').toString();
+  }
+  const query = queryStringToQuery(request.querystring);
+  const signed = await sigv4.sign({
+    method: request.method,
+    headers: headerBag,
+    hostname: headerBag.host,
+    path: request.uri,
+    body,
+    query,
+    protocol: 'https',
+  });
+  request.headers = headerBagToCfHeaders(signed.headers);
+}
+
+function getSigV4(): SignatureV4 {
+  const region = process.env.AWS_REGION;
+  const accessKeyId = process.env.AWS_ACCESS_KEY_ID;
+  const secretAccessKey = process.env.AWS_SECRET_ACCESS_KEY;
+  const sessionToken = process.env.AWS_SESSION_TOKEN;
```

Comment:
i thought edge functions didn't have environment variables, are these automatically provided? i can't find docs on them

Reply:
You're correct that environment variables cannot be used within Lambda@Edge functions. However, `process.env` is still injected with AWS credentials per request such that the AWS SDK's can be used within the functions to make API calls to AWS.

Reply:
if it works then 👍🏻 
i just couldn't find any docs on what env vars edge functions have available